client/db/bolt/upgrades.go

```diff
@@ -84,16 +84,28 @@ func (db *BoltDB) upgradeDB() error {
 		return fmt.Errorf("failed to backup DB prior to upgrade: %w", err)
 	}
 
-	return db.Update(func(tx *bbolt.Tx) error {
-		// Execute all necessary upgrades in order.
-		for i, upgrade := range upgrades[version:] {
-			err := doUpgrade(tx, upgrade, version+uint32(i)+1)
-			if err != nil {
-				return err
-			}
+	// Each upgrade its own tx.
+	for i, upgrade := range upgrades[version:] {
+		err = db.Update(func(tx *bbolt.Tx) error {
+			return doUpgrade(tx, upgrade, version+uint32(i)+1)
+		})
+		if err != nil {
+			return err
 		}
-		return nil
-	})
+	}
+	return nil
+
+	// All upgrades in a single tx.
+	// return db.Update(func(tx *bbolt.Tx) error {
+	// 	// Execute all necessary upgrades in order.
+	// 	for i, upgrade := range upgrades[version:] {
+	// 		err := doUpgrade(tx, upgrade, version+uint32(i)+1)
+	// 		if err != nil {
+	// 			return err
+	// 		}
+	// 	}
+	// 	return nil
+	// })
```

Comment:
Will remove this commented chunk.  I feel like it's not horrible if a rollback only goes to one version up instead of all the way to the starting version because at least the DB isn't hosed.  It does have an impact on memory, so I don't think we have a choice.  See the `TestUpgradeDB` test with the dexbot-testnet.db.gz archive.

All in 1 txn:

```
File: bolt.test
Type: alloc_space
Time: Apr 27, 2021 at 2:41pm (CDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 3861.84MB, 92.23% of 4187.32MB total
Dropped 8 nodes (cum <= 20.94MB)
Showing top 10 nodes out of 45
      flat  flat%   sum%        cum   cum%
  673.61MB 16.09% 16.09%   673.61MB 16.09%  go.etcd.io/bbolt.Open.func1
  583.08MB 13.92% 30.01%   583.08MB 13.92%  go.etcd.io/bbolt.(*node).dereference
  464.23MB 11.09% 41.10%   464.23MB 11.09%  go.etcd.io/bbolt.(*node).put
  451.83MB 10.79% 51.89%   451.83MB 10.79%  go.etcd.io/bbolt.(*node).read
  442.72MB 10.57% 62.46%   442.72MB 10.57%  decred.org/dcrdex/dex/encode.ExtractPushes
  407.78MB  9.74% 72.20%   407.78MB  9.74%  go.etcd.io/bbolt.(*Bucket).write
  314.68MB  7.52% 79.72%   369.18MB  8.82%  go.etcd.io/bbolt.(*Bucket).openBucket
  211.51MB  5.05% 84.77%   211.51MB  5.05%  go.etcd.io/bbolt.(*Cursor).search
  165.38MB  3.95% 88.72%   617.21MB 14.74%  go.etcd.io/bbolt.(*Bucket).node
  147.02MB  3.51% 92.23%   147.02MB  3.51%  decred.org/dcrdex/dex/encode.BuildyBytes.AddData
```

Separate txns:

```
File: bolt.test
Type: alloc_space
Time: Apr 27, 2021 at 2:38pm (CDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 3212.76MB, 91.10% of 3526.49MB total
Dropped 24 nodes (cum <= 17.63MB)
Showing top 10 nodes out of 45
      flat  flat%   sum%        cum   cum%
  453.32MB 12.85% 12.85%   453.32MB 12.85%  go.etcd.io/bbolt.(*node).read
  435.18MB 12.34% 25.20%   435.18MB 12.34%  go.etcd.io/bbolt.(*node).put
  420.21MB 11.92% 37.11%   420.21MB 11.92%  decred.org/dcrdex/dex/encode.ExtractPushes
  398.27MB 11.29% 48.40%   398.27MB 11.29%  go.etcd.io/bbolt.(*Bucket).write
  364.41MB 10.33% 58.74%   364.41MB 10.33%  go.etcd.io/bbolt.Open.func1
  318.06MB  9.02% 67.76%   318.06MB  9.02%  go.etcd.io/bbolt.(*node).dereference
  311.17MB  8.82% 76.58%   356.18MB 10.10%  go.etcd.io/bbolt.(*Bucket).openBucket
  201.51MB  5.71% 82.30%   201.51MB  5.71%  go.etcd.io/bbolt.(*Cursor).search
  174.11MB  4.94% 87.23%   627.43MB 17.79%  go.etcd.io/bbolt.(*Bucket).node
  136.51MB  3.87% 91.10%   136.51MB  3.87%  decred.org/dcrdex/dex/encode.BuildyBytes.AddData
```

Reply:
With the current changes:

```
top
Showing nodes accounting for 3515.76MB, 91.35% of 3848.73MB total
Dropped 16 nodes (cum <= 19.24MB)
Showing top 10 nodes out of 45
      flat  flat%   sum%        cum   cum%
  648.01MB 16.84% 16.84%   648.01MB 16.84%  go.etcd.io/bbolt.Open.func1
  463.85MB 12.05% 28.89%   463.85MB 12.05%  go.etcd.io/bbolt.(*node).read
  451.74MB 11.74% 40.63%   451.74MB 11.74%  go.etcd.io/bbolt.(*node).put
  444.22MB 11.54% 52.17%   444.22MB 11.54%  decred.org/dcrdex/dex/encode.ExtractPushes
  381.26MB  9.91% 62.07%   381.26MB  9.91%  go.etcd.io/bbolt.(*Bucket).write
  319.06MB  8.29% 70.36%   319.06MB  8.29%  go.etcd.io/bbolt.(*node).dereference
  273.66MB  7.11% 77.47%   330.16MB  8.58%  go.etcd.io/bbolt.(*Bucket).openBucket
  206.01MB  5.35% 82.83%   206.01MB  5.35%  go.etcd.io/bbolt.(*Cursor).search
  172.94MB  4.49% 87.32%   636.78MB 16.55%  go.etcd.io/bbolt.(*Bucket).node
  155.02MB  4.03% 91.35%   155.02MB  4.03%  decred.org/dcrdex/dex/encode.BuildyBytes.AddData
```

Your second top is missing `go.etcd.io/bbolt.Open.func1`, wonder why?

Reply:
I wonder why too. That profile looks more like the profile with a single Update as it was before.  :confused:

Reply:
With https://github.com/decred/dcrdex/pull/1070/commits/d977162eb145d4276e3df85c81a81d5ca708b5e4, now I'm seeing:

```
$  go test -v -count 1 -run TestUpgradeDB -memprofile=mem.out
=== RUN   TestUpgradeDB
2021-04-28 11:58:11.281 [INF] db_TEST: Upgrading database from version 1 to 3
2021-04-28 11:58:11.513 [DBG] db_TEST: Upgrading to version 2...
2021-04-28 11:58:15.132 [DBG] db_TEST: Upgrading to version 3...
--- PASS: TestUpgradeDB (6.90s)
PASS
ok  	decred.org/dcrdex/client/db/bolt	7.221s
```

```
$  go tool pprof bolt.test mem.out
File: bolt.test
Type: alloc_space
Time: Apr 28, 2021 at 11:58am (CDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top15
Showing nodes accounting for 2627.65MB, 99.25% of 2647.48MB total
Dropped 14 nodes (cum <= 13.24MB)
Showing top 15 nodes out of 43
      flat  flat%   sum%        cum   cum%
  458.21MB 17.31% 17.31%   458.21MB 17.31%  go.etcd.io/bbolt.(*node).put
  364.92MB 13.78% 31.09%   364.92MB 13.78%  go.etcd.io/bbolt.Open.func1
  335.56MB 12.67% 43.77%   335.56MB 12.67%  go.etcd.io/bbolt.(*node).dereference
  298.68MB 11.28% 55.05%   361.68MB 13.66%  go.etcd.io/bbolt.(*Bucket).openBucket
  247.69MB  9.36% 64.40%   247.69MB  9.36%  go.etcd.io/bbolt.(*node).read
  216.66MB  8.18% 72.59%   216.66MB  8.18%  go.etcd.io/bbolt.(*Bucket).write
  163.01MB  6.16% 78.74%   163.01MB  6.16%  go.etcd.io/bbolt.(*Cursor).search
  154.58MB  5.84% 84.58%   154.58MB  5.84%  decred.org/dcrdex/dex/encode.ExtractPushes
  127.05MB  4.80% 89.38%   127.05MB  4.80%  decred.org/dcrdex/client/db.decodeMatchProof_v2
   92.26MB  3.49% 92.87%   339.96MB 12.84%  go.etcd.io/bbolt.(*Bucket).node
   81.67MB  3.08% 95.95%   537.36MB 20.30%  go.etcd.io/bbolt.(*Bucket).Bucket
      63MB  2.38% 98.33%       63MB  2.38%  go.etcd.io/bbolt.newBucket (inline)
      10MB  0.38% 98.71%  1049.60MB 39.65%  go.etcd.io/bbolt.(*Bucket).spill
    7.50MB  0.28% 98.99%   759.83MB 28.70%  go.etcd.io/bbolt.(*Bucket).Put
    6.86MB  0.26% 99.25%   709.17MB 26.79%  go.etcd.io/bbolt.(*Tx).allocate

```